esp-hal/src/i2c/master/mod.rs

```diff
@@ -763,11 +763,10 @@ impl core::future::Future for I2cFuture<'_> {
         self.state.waker.register(ctx.waker());
```

Comment:
s.th. like this 
```rust
        if self.is_done() {
            Poll::Ready(Ok(()))
        } else if self.check_error().is_err() {
            Poll::Ready(error)
        } else {
            self.state.waker.register(ctx.waker());
            Poll::Pending
        }
```

would avoid doing anything else before checking done

Reply:
It's also a race condition, unfortunately:

If we are interrupted after is_done returns false, the hardware can complete and run into a timeout as before this PR, and then when the Future regains control it'll return an error even though the transfer finished correctly.

You may be right about the waker part, I'm not entirely sure currently.

Reply:
Since INT_ENA isn't being enabled/disabled here, you have to register the waker before checking.